massa-versioning-worker/src/versioning.rs

```diff
@@ -0,0 +1,1057 @@
+    /// Create
+    pub fn new(defined: MassaTime) -> Self {
+        let state: MipState = Default::default();
+        let state_id = MipStateTypeId::from(&state);
+        // let mut advance = Advance::default();
+        // advance.now = defined;
+        let advance = Advance {
+            now: defined,
+            ..Default::default()
+        };
+        let history = BTreeMap::from([(advance, state_id)]);
+        Self {
+            state: Default::default(),
+            history,
+        }
+    }
+
+    /// Advance state
+    pub fn on_advance(&mut self, input: &Advance) {
+        let now = input.now;
+        // Check that input.now is not after last item in history
+        // We don't want to go backward
+        let is_forward = self
+            .history
+            .last_key_value()
+            .map(|(adv, _)| adv.now < now)
+            .unwrap_or(false);
+
+        if is_forward {
+            // machines crate (for state machine) does not support passing ref :-/
+            let state = self.state.on_advance(input.clone());
+            // Update history as well
+            if state != self.state {
+                let state_id = MipStateTypeId::from(&state);
+                self.history.insert(input.clone(), state_id);
+                self.state = state;
+            }
+        }
+    }
+
+    /// Given a corresponding VersioningInfo, check if state is coherent
+    /// it is coherent
+    ///   if state can be at this position (e.g. can it be at state "Started" according to given time range)
+    ///   if history is coherent with current state
+    /// Return false for state == MipState::Error
+    pub fn is_coherent_with(&self, versioning_info: &MipInfo) -> bool {
+        // Always return false for state Error or if history is empty
+        if matches!(&self.state, &MipState::Error) || self.history.is_empty() {
+            return false;
+        }
+
+        // safe to unwrap (already tested if empty or not)
+        let (initial_ts, initial_state_id) = self.history.first_key_value().unwrap();
+        if *initial_state_id != MipStateTypeId::Defined {
+            // self.history does not start with Defined -> (always) false
+            return false;
+        }
+
+        // Build a new VersionStateHistory from initial state, replaying the whole history
+        // but with given versioning info then compare
+        let mut vsh = MipStateHistory::new(initial_ts.now);
+        let mut advance_msg = Advance {
+            start_timestamp: versioning_info.start,
+            timeout: versioning_info.timeout,
+            threshold: Amount::zero(),
+            now: initial_ts.now,
+        };
+
+        for (adv, _state) in self.history.iter().skip(1) {
+            advance_msg.now = adv.now;
+            advance_msg.threshold = adv.threshold;
+            vsh.on_advance(&advance_msg);
+        }
+
+        // XXX: is there always full eq? Can we have slight variation here?
+        vsh == *self
```

Comment:
I think this is correct, because you recreate the only things that should evolve in the Advance message (now and threshold)

Reply:
Removed the XXX comment